node/impl/client/import.go

```diff
@@ -2,75 +2,96 @@ package client
 	"golang.org/x/xerrors"
 )
 
-// importNormalFileToFilestoreCARv2 transforms the client's "normal file" to a Unixfs IPLD DAG and writes out the DAG to a CARv2 file
-// that can be used to back a filestore.
-func (a *API) importNormalFileToFilestoreCARv2(ctx context.Context, importID importmgr.ImportID, inputFilePath string, outputCARv2Path string) (c cid.Cid, finalErr error) {
-
-	// TODO: We've currently put in a hack to create the Unixfs DAG as a CARv2 without using Badger.
-	// We first create the Unixfs DAG using a filestore to get the root of the Unixfs DAG.
-	// We can't create the UnixfsDAG right away using a CARv2 read-write blockstore as the blockstore
-	// needs the root of the DAG during instantiation to write out a valid CARv2 file.
+// doImport takes a standard file (src), forms a UnixFS DAG, and writes a
+// CARv2 file with positional mapping (backed by the go-filestore library).
+func (a *API) doImport(ctx context.Context, src string, dst string) (cid.Cid, error) {
+	// This method uses a two-phase approach with a staging CAR blockstore and
+	// a final CAR blockstore.
+	//
+	// This is necessary because of https://github.com/ipld/go-car/issues/196
 	//
-	// In the second pass, we create a CARv2 file with the root present using the root node we get in the above step.
-	// This hack should be fixed when CARv2 allows specifying the root AFTER finishing the CARv2 streaming write.
-	fm := filestore.NewFileManager(ds_sync.MutexWrap(datastore.NewMapDatastore()), "/")
-	fm.AllowFiles = true
-	fstore := filestore.NewFilestore(bstore.NewMemorySync(), fm)
+	// TODO: do we need to chunk twice? Isn't the first output already in the
+	//  right order? Can't we just copy the CAR file and replace the header?
+	f, err := ioutil.TempFile("", "")
+	if err != nil {
+		return cid.Undef, xerrors.Errorf("failed to create temp file: %w", err)
+	}
+	_ = f.Close() // close; we only want the path.
+	tmp := f.Name()
+	defer os.Remove(tmp)
+
+	// Step 1. Compute the UnixFS DAG and write it to a CARv2 file to get
+	// the root CID of the DAG.
+	fstore, err := stores.ReadWriteFilestore(tmp)
+	if err != nil {
+		return cid.Undef, xerrors.Errorf("failed to create temporary filestore: %w", err)
+	}
+
 	bsvc := blockservice.New(fstore, offline.Exchange(fstore))
-	defer bsvc.Close() //nolint:errcheck
+	dags := merkledag.NewDAGService(bsvc)
 
-	// ---- First Pass ---  Write out the UnixFS DAG to a rootless CARv2 file by instantiating a read-write CARv2 blockstore without the root.
-	root, err := importNormalFileToUnixfsDAG(ctx, inputFilePath, merkledag.NewDAGService(bsvc))
+	root, err := buildUnixFS(ctx, src, dags)
 	if err != nil {
-		return cid.Undef, xerrors.Errorf("failed to import file to store: %w", err)
+		_ = fstore.Close()
+		return cid.Undef, xerrors.Errorf("failed to import file to store to compute root: %w", err)
+	}
+
+	if err := fstore.Close(); err != nil {
+		return cid.Undef, xerrors.Errorf("failed to finalize car filestore: %w", err)
```

Comment:
Should this be fstore.Finalize()?

Reply:
Nope, because we use `stores.ReadWriteFilestore()` which returns a ClosableBlockstore whose `Close` method calls `Finalize`. This design was already in place in the wip branch, we certainly can review it because it's rather footgunny.

(In a previous review of go-fil-markets I had removed these utilities because they created obscurity IMO -- see commit [`4d3607a` (#596)](https://github.com/filecoin-project/go-fil-markets/pull/596/commits/4d3607a44c8d9f7ce14def8d21bf2436fcb5cff9), but then readded them to keep code DRY -- happy to revisit!)